audit_log/admin.py

```diff
@@ -0,0 +1,228 @@
+        message = create_api_commit_message_from_request(
+            request=request,
+            operation=Operation.UPDATE.value if change else Operation.CREATE.value,
+            object_ids=[str(obj.pk)],
+            model_name=obj._meta.model_name,
+        )
+        audit_log_service._commit_to_audit_log(message=message)
+        super().save_model(request, obj, form, change)
+
+    def delete_model(self, request, obj):
+        """
+        Given a model instance delete it from the database.
+        """
+        message = create_api_commit_message_from_request(
+            request=request,
+            operation=Operation.DELETE.value,
+            object_ids=[str(obj.pk)],
+            model_name=obj._meta.model_name,
+        )
+        audit_log_service._commit_to_audit_log(message=message)
+        super().delete_model(request, obj)
+
+    def delete_queryset(self, request, queryset):
+        """Given a queryset, delete it from the database."""
+        super().delete_queryset(
+            request,
+            queryset.with_audit_log_and_request(
+                request=request, operation=Operation.DELETE.value
+            ),
+        )
+
+
+class AuditLogMessageOperationListFilter(admin.SimpleListFilter):
+    # Human-readable title which will be displayed in the
+    # right admin sidebar just above the filter options.
+    title = _("operation")
+
+    # Parameter for the filter that will be used in the URL query.
+    parameter_name = "operation"
+
+    options = [operation.value for operation in Operation]
+
+    def lookups(self, request, model_admin):
+        return map(lambda o: (o, o.upper()), self.options)
+
+    def queryset(self, request, queryset):
+        if self.value() in self.options:
+            return queryset.filter(message__icontains=self.value())
+        return queryset
+
+
+class AuditLogMessageRoleListFilter(admin.SimpleListFilter):
+    # Human-readable title which will be displayed in the
+    # right admin sidebar just above the filter options.
+    title = _("role")
+
+    # Parameter for the filter that will be used in the URL query.
+    parameter_name = "role"
+
+    options = [role.value for role in Role]
+
+    def lookups(self, request, model_admin):
+        return map(lambda o: (o, o.upper()), self.options)
+
+    def queryset(self, request, queryset):
+        if self.value() in self.options:
+            return queryset.filter(message__icontains=self.value())
+        return queryset
+
+
+class LargeTablePaginator(Paginator):
+    """
+    Paginator that uses PostgreSQL `reltuples` for queryset size.
+
+    The reltuples values is just an estimate, but it's much faster than
+    the COUNT(*) which is used by the default paginator. Therefore, this
+    should work better for tables containing millions of rows.
+
+    See https://wiki.postgresql.org/wiki/Count_estimate for details.
+    """
+
+    @cached_property
+    def count(self):
```

Comment:
According to Python [documentation](https://docs.python.org/3/library/functools.html#functools.cached_property) this calculates the value and caches it for the lifetime of the instance. So, when is the instance created? Is the cached value correct or should it change during the instance's lifetime?

Reply:
`@cached_property` is also used in [the original implementation in Django](https://github.com/django/django/blob/main/django/core/paginator.py#L106), so there is absolutely nothing wrong in using it.